### DISCUSSION TITLE
Alternative mutation rework proposal

### DISCUSSION THREAD

**Discussion by Venera3** [CleverRaven/Cataclysm-DDA#48086] [Ideas]
With the release of the 0.F Stable and with that the next experimental cycle approaching I wanted to break out my soapbox and pitch an alternative to #28277, since whichever way the rework goes I'd like to contribute a good bit to the content side of it. This is for the most part based on repurposing existing systems, but my usual "I'm not a coder" caveat applies.

### Goals

Mostly the same as those of Erk's proposal: 

- Turn mutation into a longer-term process both to draw out the midgame and to discourage savescumming
- Declutter mutagen crafting
- Allow for a more narrow specialization, if you can source the materials
- Do away with the hard mutagen/serum divide and allow for alternative mutation sources (with their own drawbacks)

### How it would work

Instead of the mostly strict mutation categories each mutation has any number of `aspect`s defined, these each correspond to a new type of vitamin `mutagen`. Each mutation also lists a `mutagen_cost`, a number of `mutation_types`, `prerequisites`, and `body_parts` it can be applied to.

Instead of crafting a purpose-made mutagen you craft a generic recipe for which you can select from a long list of ingredients, each contributing different `mutagen` content to the finished product. The difference between "off-brand" mutation sources, normal mutagens, and serums would only be the composition and strength of the mutagen content in the result:

- Normal mutagen results in a 50/50 mix of `mutagen_generic` and whichever `mutagen_aspect` the ingredients had,
- Serum applies twice the `mutagen_aspects` of its ingredients and no generic mutagen aspect
- Suboptimal mutagenic sources (nether influence, chewing on random bits in labs) apply a special `mutagen_bad` aspect that's assigned to all negative mutations irrespective of their other aspects

Applying mutagenic sources accumulates their `mutagen` content,  when the combined amount of them reaches a set point it applies a "Phenopotency" effect (yoinked that over for clarity). This has a set chance to mutate the survivor every X hours, with increasing `mutagen` levels reducing the time between triggering the roll. This roll depends on two things: the combined amount of mutagens in your system (`effective mutagen`) and the amount of each discrete `mutagen`. 

- First, all mutations with `mutation_cost` above your `effective mutagen` are eliminated
- Then a roll weighted by the proportion of different mutagens selects a mutation from the remainder and checks if 
```(its mutagen_cost`) - (mutagen_cost of its prerequisites) + (mutagen_cost of conflicting traits)``` 
is equal to or smaller than your sum of relevant mutagens
- If you can afford it  the mutation is now designated as a mutation goal and its mutagen cost is subtracted from your relevant mutagen levels. If it's higher you get a failure message and it rolls again from scratch at the next try
- The roll's exact results are hidden from the player, they instead get a random message depending on the body part/mutation type of the designated mutation to hint at what's coming
- After a set amount of time, possibly depending on the mutagen cost of the trait the characters develop an intermediary trait (preferably also dynamic based on body part/type) that may be positive, negative or mixed, this trait will turn into the designated mutation goal after another dormant period

Mutations are roughly tiered:
- A good number of flavor/body horror mutations that can use any aspect and have a low mutagen cost
- Basic mutations have `mutagen_generic` as one of their aspects, increasing their chance to be affordable even on entry-level mutagens
- Slightly more specialized mutations require broader aspects mapping to 2-3 combined categories of the current system
- Specialized mutations require category-equivalent aspects
- Threshold mutations require high mutagen costs of their aspect, but in addition to gating the high-level mutations they also increase your intrinsic level of the relevant aspect, making staying in your category more likely
- Some posttresh mutations are extremely specialized, having only one particular aspect but offering major boosts
- Most mutations have one or more overblown levels that are either actively harmful or can cause the player to mutate uncontrollably (either by producing `mutagen_generic` or `mutagen_bad`)

Mutation conflicts are automatic instead of being defined by hand, with `body_parts` (arm, torso, whole body etc) determining locational and `mutation_type` (mental, skin, metabolism, threshold etc) determining other conflicts. Categories are also JSON-definable along with their conflict conditions ( always exclusive, body part exclusive, non-conflicting).

Purifier removes a given amount of mutagens and accumulates its own vitamin type. When this reaches a threshold an equal effect to "Phenopotency" is applied, with a roll similar to that to determine which mutation it attempts to remove, weighted roughly 
```last mutation > most expensive mutation > rest in decreasing base cost >>>starting traits```
Removing traits also takes time and goes through some potentially debilitating intermediate stages and costs the base `mutagen_cost` in purifier. 

Both purifiers and mutagens have a separate long-term system akin to toxins to discourage binging them and to nudge players towards sourcing the more concentrated versions. Both also have a natural decay, further discouraging metagaming their effects (see the example below).

**Result**

- The mutagen ingredients you use influence the resulting mutations, and allow for more specific direction even at a relatively low-tech level (but have fun sourcing bear marrow for it)
- The aspects are more gradual than categories, allowing for some out-there mutations instead of being pretty much category-locked after a few doses
- Thresholds are *technically* not gated behind serums, but reaching them with basic mutagens is unlikely 
- Longer mutation/removal timeline keeps it from being easily optimized
- Automatic conflicting removes the potential for oversights
- Mutation sources, mutagen strength, and the timeline of the whole system are JSON-driven, allowing for easy modding/fixing down the line

### Example

Let's say that basic mutagen turns half of the ingredient's aspects into generic mutagen, centipede eggs as an ingredient provide 100 arthropod and 100 centipede aspect, natural rate of mutagen decays is 5/h, Phenopotency has a 50%/h chance to roll for a mutation and there are three chitin mutation variants and a common precursor mutation:
```
"id": "pre_chitin",
"aspects": [ "ARTHROPOD", "INSECT", "SPIDER", "WASP", "BEE", "ANT", "CENTIPEDE", "GENERIC" ],
"mutagen_cost": 90,
"body_parts": [ "all" ],
"category": [ "SKIN" ]

"id": "chitin_furred",
"aspects": [ "SPIDER", "BEE" ],
"prereq": [ "pre_chitin" ],
"mutagen_cost": 300

"id": "chitin_plate",
"aspects": [ "ANT", "CENTIPEDE" ],
"mutagen_cost": 400

"id": "chitin_light",
"aspects": [ "INSECT", "WASP", "ARTHROPOD" ],
"mutagen_cost": 150
```
- The player drinks their mutagen, landing them at 100 generic, 50 arthropod and 50 centipede mutagen and activating Phenopotency
- After an hour the mutagens decayed to 95 generic, 45 arthropod and 45 centipede, Phenopotency succeeds in the chance to mutate roll
- All mutations above 185 mutagen_cost are discarded
- The mutation roll is weighted (essentially) 50% generic 25% arthropod 25% centipede
- Luckily we rolled pre_chitin, it's now a designated mutation target and the player gets a message about their whole body itching slightly. All three mutagens are reduced by a third of the mutagen_cost, resulting in 65 generic, 15 arthropod and 15 centipede mutagen
- Another hour passes, mutagens decrease to 60/10/10, this chance to mutate roll is successful
- All mutations above 80 cost are discarded, the roll to choose is weighted 75%/12,5%/12,5%
- We roll Meat Preference, a low-level centipede/spider/etc mutation that costs 50
- We can't pay the cost, message the player about a failure 
- The next two chance to mutate rolls fail, the third gives our player a minor flavor mutation (Creaking joints, no effect on its own but might turn into Set Joints if exposed to `mutagen_bad`), taking us below the level necessary for Phenopotency
- In the meantime the timer for our intermediate mutation ticked down and our player now gets to enjoy Burning Skin on their whole body, causing morale loss and periodic scratching. After another few hours it happily turns into Chitinous Patches, providing minor cut/pierce resistance and opening the door to the higher chitin variants.

**Comment by actual-nh** [comment]
Nice!

Being able to have things not quite as on-off - aspects having variable strengths (in mutations as well as in tendencies toward), for instance - would be an attractive potential feature of this system, with not being as locked-in to a particular pathway. (That also fits rather better with quite a few of the mutated animals - what pathway is that 7-legged, 5-eyed (or whatever) deer on?) Note that this being non-binary could also be so for the conflicting determination - some mutations are more difficult to have together (a whole bunch of digestive system ones, say) but not impossible. Would this have any removal of contradictory traits other than via purifier? I'd say that could be done with a system of "decreased likelihood but not zilch of getting a conflicting mutation that wipes out the trait".

This would also be an opportunity to get a "general+specific" framework set up, that could be useful for things like anger/fear (aggression/morale) - yes, the critter initially is irked/scared in general, but that goes away, while the memory of whatever specifically did it remains a lot longer.

**Reply by Venera3** [reply]
Mutating towards a conflicting trait would be made less likely by their increased cost, but that would be easy to overcome with ingesting lots of the stuff. That would mean you'd need to ingest small amounts to stay on your existing mutation's tree(since your mutation prereqs mean their follow-up mutations are cheaper) but if you wanted something completely different you could raise its chances by going overboard.

That ties to how prereqs are handled in general, which I'm still a bit torn on. I originally sketched out the example above with the assumption that in-progress prerequisite mutations are counted for the cost, so that you could "upgrade" it before it manifests properly - in the example, ingesting another mutagen allowing you at a roll for the other variants *if* pre_chitin was chosen. Another option is that they aren't hard prerequisites at all, only making the follow-up mutations cheaper so that drinking 5 mutagens in the example could roll chitin_plate directly, but it's less likely than if you took it slow and waited out the intermediate mutations.

**Comment by Aut-Baron** [comment]
Also, I believe that if I'm in the cataclysm, eventually I'll learn how to better use my mutations and may make specialized styles/ ways of using the mutation depending on the mutation itself. So, I propose that using a mutation enough gives you a "mutation"(it'll show up on the active mutation list) that you can use and do something like with... Vine Sprouter mutation and  maybe hair roots. At the cost of a lot of energy you could rapidly sprout and detach vines, and I mean rapidly, in an attempt to (like a panic button) make a large burst of vines outward or maybe even just extending the vines(which should be a thing in game anyways but, if someone knows how that might mess up the code then that's probably the reason it's not in here.). Heck, with shell you could do something like a shell slide, in which, you dive in a direction and retract into your shell at the same time in an attempt to knock down your enemies our just get away quickly. In all honesty, with enough dodging skill(because its described as helping in acrobatic feats) you should be able to just straight up dodge roll or dive out of the way of an incoming Juggernaut stomp and move yourself like maybe 1-3 tiles.
So, in summary:

- Make using mutations give you active mutations or passive mutations (like crafty, where it says you learn how to take advantage of your size) based on the individual mutations themselves.
- Maybe(NOT SUGGESTED ABOVE) make a martial arts learned automatically at a certain number in both melee and unarmed skill
- Make a couple things make sense(like a tail slap with a fin should be poosible even though it wouldnt be strong)

**Comment by Altogolik** [comment]
I would like to be able to make mutations instant, I often save and load and the system with a slow mutation may break and I will not get any mutations at all (it broke my character once)
+ I don't like to wait for the result, I hate a new mutation system.
